core/services/workflows/engine.go

```diff
@@ -31,6 +31,52 @@ type stepRequest struct {
 	state   store.WorkflowExecution
 }
 
+type stepUpdateChannel struct {
+	executionID string
+	ch          chan store.WorkflowExecutionStep
+}
+
+type stepUpdateManager struct {
+	mu sync.RWMutex
+	m  map[string]stepUpdateChannel
+}
+
+func (sucm *stepUpdateManager) add(executionID string, ch stepUpdateChannel) (added bool) {
+	sucm.mu.Lock()
+	defer sucm.mu.Unlock()
+	if _, ok := sucm.m[executionID]; ok {
+		return false
+	}
```

Comment:
Could add a fast read path here:
```suggestion
	sucm.mu.RLock()
	_, ok := sucm.m[executionID]
	sucm.mu.RUnlock()
	if ok {	
		return false
	}
	sucm.mu.Lock()
	defer sucm.mu.Unlock()
	if _, ok := sucm.m[executionID]; ok {
		return false
	}
```